integrate coredev.buildout (was PR #1670) (#1671)
plone/documentation: docs/contributing/core/index.md

    "property=og:title": "Contribute to Plone 6 Core"
    "keywords": "Plone, Plone Contributor Agreement, License"
---

# Contribute to Plone 6 core

This guide describes the process of how to contribute to, and develop in, Plone core.
It expands upon {doc}`/contributing/index`.

This guide assumes that you have basic knowledge of how to use git and GitHub.
If you have never contributed to Plone, or you lack basic knowledge of how to use git and GitHub, you should first read {doc}`/contributing/first-time` for more information.

```{important}
You must {ref}`contributing-sign-and-return-the-plone-contributor-agreement-label` before your contribution can be accepted.
```

## Version support policy

Before you contribute to Plone core, check the [version support policy](https://plone.org/download/release-schedule) to see which versions of Plone are currently supported.

## Pre-requisites

It is beyond the scope of this documentation to provide installation instructions for all pre-requisites for your operating system.
However, the following links and sections below may be helpful.

```{include} ../../volto/contributing/install-operating-system.md
```

> -   Python {SUPPORTED_PYTHON_VERSIONS}

Reviewer: This is version-dependent.

Plone 6.1 will only support Python 3.10 through 3.12 (see https://community.plone.org/t/plone-6-1-will-drop-support-for-python-3-8-and-3-9/19128)

Plone 6.0 supports Python 3.8 through 3.12

Plone 5.2 supports Python 2.7 and 3.8

Author: I could replace the replacement with Plone-specific versions when 6.1 is released.

- PLONE_61_SUPPORTED_PYTHON_VERSIONS
- PLONE_60_SUPPORTED_PYTHON_VERSIONS

Plone 5.2 development is documented at 5.docs.plone.org.

Author: See https://github.com/plone/documentation/issues/1674 and https://github.com/plone/Products.CMFPlone/issues/3926

-   {term}`GNU make`
-   {term}`Git`
-   [libxml2 and libxslt](https://gitlab.gnome.org/GNOME/libxslt/-/releases), including development headers

> -   [GNU Compiler Collection (GCC)](https://gcc.gnu.org/) to compile {term}`ZODB`, {term}`Zope`, and {term}`lxml`

Reviewer: You need a C compiler. It can be gcc, or clang (which is what the macOS developer tools give you)

Author: See ef40380271ffed73f2f45f052203193add919fdf

### Python

Installing Python is beyond the scope of this documentation.
However, it is recommended to use a Python version manager, {term}`pyenv` that allows you to install multiple versions of Python on your development environment without destroying your system's Python.
Plone requires Python version {SUPPORTED_PYTHON_VERSIONS}.

### Make

```{include} ../../volto/contributing/install-make.md
```

### Git

```{include} ../../volto/contributing/install-git.md
```

### libxml2, libxslt, and GCC

Consult the resources at the links above for installation instructions.

> Package managers may facilitate installation.

Reviewer: These days on most platforms lxml can be installed as a prebuilt binary wheel that includes statically linked copies of libxml2 and libxslt. So in most cases this doesn't need to be done manually.

I think we could probably get away with removing it from the list of prerequisites here. We already don't mention it in https://plone6--1671.org.readthedocs.build/install/create-project.html

Author: See ef40380271ffed73f2f45f052203193add919fdf

## Install Plone core for development

To set up a Plone 6 development environment, change your working directory to wherever you place your projects, and clone https://github.com/plone/buildout.coredev.

```shell
cd [MY_PROJECTS]
git clone https://github.com/plone/buildout.coredev
cd buildout.coredev
```

Now run the script to install Plone 6.

```shell
./bootstrap.sh

> ```

Reviewer: Note, this script currently requires python3.11 specifically to be available in user's PATH

Author: There's also `.bootstrap-pyenv.sh`, which appears to be targeted toward Plone 6.1, so I omitted it.

I added a conditional step in cbb149e610f953e3f468ca2a2f844ff51b853fc0.

If you run into issues in this process, see {doc}`troubleshooting`.

This will run for a long time if it's your first pull (approximately 10-20 minutes, depending on network speed and your hardware).

Once that's done, you can start an instance of Plone with the following command.

```shell
./bin/instance fg
```

> To visit your Plone instance, you can open the link http://0.0.0.0:8080 in a web browser.

Reviewer: We should clarify that this is the way to access the Classic UI. If you want to use Volto with the backend installed by the coredev buildout, then you have to:
1. Run the coredev buildout, start it, and create a Volto site.
2. Install and run Volto, and access it on a different port.

Author: Just to be clear, the buildout method has links to install Volto as you suggest in Step 2.

![Screenshot 2024-06-24 at 3 11 43 AM](https://github.com/plone/documentation/assets/102112/eef337fb-d3ea-4fe1-810c-2924ca1e57f1)

[Install Plone from its packages](https://6.docs.plone.org/install/install-from-packages.html) and refer to the full Volto documentation [Frontend](https://6.docs.plone.org/volto/index.html).

Both links are incorrect.

We want to [Develop Volto core](https://6.docs.plone.org/volto/contributing/developing-core.html). But even that is not correct because the backend is in a Docker container.

So how does one install and run Volto in this context?

Until there exists a clear way to do this, I will leave Volto out of the coredev documentation.

Addressed in [edee66e](https://github.com/plone/documentation/pull/1671/commits/edee66e4fdb4df1fe47956ea57b8691db2a17887).

Author: Just saw this come through the daily flood of GitHub emails:

https://github.com/plone/plone.volto/issues/155

@ericof @davisagli @sneridagh is `plone.volto`'s `README.md` the official docs for how to install Volto using buildout? The missing link, if you will.

Reviewer: @stevepiercy The change you made is good enough for now.

We could add a note to https://6.docs.plone.org/volto/contributing/developing-core.html#start-the-backend-and-volto explaining that you also have the option of installing and starting the Plone backend in some other way -- such as with the coredev buildout instead of `make start-backend-docker`. As long as it's running on port 8080, the default Volto settings will use it.

If you're using buildout and you're going to use the Plone backend with Volto, then you need to choose "Create a new Plone site" instead of "Create Classic UI Plone site" on the screen that looks like this:
<img width="846" alt="Screenshot 2024-06-25 at 9 46 36 PM" src="https://github.com/plone/documentation/assets/49014/3fb22b7f-0aa8-4fd8-97bb-b3389f96ea5b">
This is what installs plone.volto which applies some extra configuration needed for Volto.

This screen will be replaced with plone.distribution's install page in Plone 6.1. If you installed Plone from the cookiecutter instead of buildout, then the make commands take care of creating the Plone site for you so you don't need to use this screen.

> Click {guilabel}`View your Plone site` to do exactly that.

Reviewer: Running the coredev buildout installs packages and creates a Zope instance, but it doesn't create a Plone site. There's a missing step here to do that before you can view the Plone site.

Author: Oops. Got ahead of myself. Fixed in 962a4b981213b3b674a6876e19e8d2d9127b01ac

To login, the default credentials are the following.

-   username: `admin`
-   password: `admin`

## Use the correct branch

```{important}
This section applies to members of the GitHub `plone/developers` team, who have write access to repositories under the Plone GitHub organization.

Members of the `plone/contributors` team do not have write access, and instead must follow the process to set up their remote upstream and origin branches as described in {ref}`set-up-your-environment-label`.
```

The current default and development branch of `buildout.coredev` is `6.1`.
Older versions are named according to their `major.minor` version.

> Always begin by checking out the git branch on which you want to work, usually called `origin`.

Reviewer: `origin` is the name of a git remote, not a branch.

Author: Thanks. I think I got it technically correct now in 7e45ba71e13dd06dddcb0afd355cb28f397ce50d

If you have not yet checked out a branch, then you need to track it with the `-t` option and specify the remote branch to track.
The following command will switch and track changes from the remote branch `origin/6.1`.

```shell
git checkout -t origin/6.1
```

If you check out a different branch, and want to return to the previous branch, you need to specify only the local branch.

```shell
git checkout 6.1
```

Next pull down and merge any recent changes from the remote tracked repository with a single command.

```shell
git pull
```

```{important}
Make sure to rerun buildout for the current branch to get the correct versions of packages, otherwise you will get some weird behavior.

> ```

Reviewer: It can be awkward to try to work on different versions of Plone in the same copy of the coredev buildout, because if you switch branches most packages will need to be reinstalled on a different version. If you need to contribute to different branches of the coredev buildout, I'd recommend creating multiple clones in different folders rather than trying to switch one clone back and forth.

Author: Addressed in 0897e15754a63afbc4266af801b0c56203f60ef5

Next create a new branch on which you want to work from the current branch, tracking the upstream Plone repository, and check it out.
It's a good idea to use a branch name that includes the issue number and is descriptive of what it resolves.

```shell

> git switch -c 123-thing-i-fixed -t origin/6.1

Reviewer: I think we want to talk about doing this in the context of a specific package that is part of Plone core, not the coredev buildout. It's very very rare to make changes to the coredev buildout itself unless you're a release manager. Usually you make a change to the package that needs it, and then once the PR for that package is merged, mr.roboto will automatically update checkouts.cfg in the coredev buildout to record the fact that that package has changes which need to be released.

I would move discussion of creating a new branch down below "Edit packages" since you have to make sure the package is checked out before you can switch it to a new branch.

Author: Addressed in 1704bbc84217f255ac9716aad06f2c0241588f89.

```

Now you can edit your code without affecting the original branch.

## Edit packages

First identify the names of the Plone packages you want to work on.
If you do not know, you can ask in the [Plone Community Forum](https://community.plone.org/).
Only a few packages are in {file}`src/` by default.

Edit the file {file}`checkouts.cfg` at the root of the repository by adding the package names under the `auto-checkout` list.

```{code-block} cfg
:emphasize-lines: 10-
[buildout]
always-checkout = force
# always-checkout = false
auto-checkout =
# These packages are always checked out:
    docs
    # <snip>
# These packages are manually added, or automatically added by mr.roboto:
    # <snip>
    plone.app.event
    icalendar
```

Then rerun buildout to install the source packages to edit.

```shell

> ./bin/buildout

Reviewer: Add the `-N` flag to save time by not checking PyPI to see if there are updates to packages that were already installed:

```suggestion
./bin/buildout -N
```

Author: Done in 7e604ed4df2c597436d9c1a1e7f8a656fb1ac31f

```

```{seealso}
For tips on working with `mr.developer`, see {doc}`mrdeveloper`.
```

## Test locally

If you change the expected behavior of a feature in a package, you should write a test to cover the change.

To run a test for the specific package that you modified, use the `-m` option followed by the package name, as shown in the following example.

```shell
./bin/test -m plone.app.caching

> ```

Reviewer: We could add a note that `bin/test` is a script installed by buildout which runs the Zope testrunner. Newcomers may be expecting pytest.

Author: That belongs more in reference, or maybe explanation, than a how-to guide. Perhaps "Test framework of Plone core"?

If any test fails, do not commit and push the changes.
Instead write a test that passes.

After the package level tests pass with your change, you can run the full unit test suite to ensure other packages aren't affected by the change.

> It takes 5-10 minutes to run the full unit test suite.

Reviewer: You can do this, but these days we have CI, so usually I create a PR at this point and ask Jenkins to run the tests.

Author: Done in fcdd9923dcb60a1007cea70e2ebf9c3bdf73544d

```shell
# Run unit tests
./bin/test
```

If you run acceptance tests with the `--all` option, it will repeatedly launch and close browser windows that gain focus, disrupting you from doing any other work.
This takes 30-40 minutes to run.
If you can't afford this disruption, you can defer it to Jenkins.

```shell
# Run acceptance tests
./bin/test --all
```

## Change log

All changes require a change log entry.

> For packages that use [towncrier](https://pypi.org/project/towncrier/) to produce change logs, see {ref}`contributing-change-log-label`.

Reviewer: It might not be obvious how to tell if it uses towncrier. Look for the `news` folder.

Author: Done in 7838b17d1b3a529800d30b4cae2d6cfd1c01d87c

For packages that don't use towncrier, edit either {file}`CHANGES.rst`, {file}`CHANGES.txt`, or {file}`HISTORY.txt` in each package you have modified, adding a summary of the change.
New change log entries should be added at the top of {file}`CHANGES.rst`.

> ## Update `checkouts.cfg`

Reviewer: Remove this section. Developers can make changes to checkouts.cfg locally in order to do development, but they should not commit them. checkouts.cfg gets updated automatically by mr.roboto after a pull request to a core package is merged.

Author: Done in abf5f7b4dadf5556cd1abe8533970eb3c083bcc6

If you didn't do so earlier, edit the file {file}`checkouts.cfg` at the root, and add your changed packages only to the `auto-checkout` list.
Remove any packages that you previously added to this file, but did not change.
Leave any pre-existing packages in the original list.

The Plone release manager will check this file to see which packages you have updated to make a proper release.

```{seealso}
{doc}`release`
```

## Create a pull request

After you have completed all the foregoing steps, push your changes to a remote branch and create a pull request in GitHub.

> If you are working from an issue, return to the original issue, and add a link to the pull request.

Reviewer: Or mention the issue as "Fixes #issuenum" in the PR description, and github will create the link

Author: Done in bc12737979848953a97e6aa2b9e18d236171b14e

## Jenkins and mr.roboto

Plone has a continuous integration ({term}`CI`) setup and follows CI rules.

When you push a change to a Plone package, there may be GitHub workflows that run automatically.
The CI package [mr.roboto](https://github.com/plone/mr.roboto) will perform some checks and suggest that you run Jenkins after all other CI runs.

> For each Plone and Python version, Jenkins runs two jobs: one for the package itself (which will give you a fast feedback, within 10 minutes) and one on the full `coredev` build (which can take up to an hour, but makes sure no other packages are affected by your change).

Reviewer: These days Jenkins does not run a job for one specific package. For most packages there are some github actions workflows that do that, as well as some lint checks.

There are usually several jobs that run all Plone tests on Jenkins -- basically one for each combination of Plone version + Python version where that branch of the package is used, but some of them may skipped to reduce resource use (i.e. for Plone 6 we run this job on Python 3.8 and 3.11 but not 3.9 or 3.10)

Author: I'm actually OK with purging this detail and not trying to document it here. It doesn't belong in a how-to guide, but more in explanation or reference.

Done in 6ef5c58872b6b68c76a8563c41a38b506ded8496

```

## Version support policy

Before you contribute to Plone core, check the [version support policy](https://plone.org/download/release-schedule) to see which versions of Plone are currently supported.

## Pre-requisites

It is beyond the scope of this documentation to provide installation instructions for all pre-requisites for your operating system.
However, the following links and sections below may be helpful.

```{include} ../../volto/contributing/install-operating-system.md
```

-   Python {SUPPORTED_PYTHON_VERSIONS}
-   {term}`GNU make`
-   {term}`Git`
-   A C compiler

### Python

Installing Python is beyond the scope of this documentation.
However, it is recommended to use a Python version manager, {term}`pyenv` that allows you to install multiple versions of Python on your development environment without destroying your system's Python.
Plone requires Python version {SUPPORTED_PYTHON_VERSIONS}.

### Make

```{include} ../../volto/contributing/install-make.md
```

### Git

```{include} ../../volto/contributing/install-git.md
```

### C compiler

> You need a C compiler on your system to compile {term}`ZODB` and {term}`Zope`.
> 
> On macOS, Developer Tools providers clang for a C compiler.

Reviewer: ```suggestion
You need a C compiler on your system to compile some of the Python libraries that Plone uses.

On macOS, Developer Tools provides Clang for a C compiler.
```

You will be redirected to your new Classic UI Plone site.

```{warning}
Ignore the warning about accessing the Plone backend through its Classic UI frontend.

Do not follow the instructions to install Volto.
They will not work with buildout.
To contribute to Volto, you will need to start over, and follow {doc}`../volto`.
```

To login, the default credentials are the following.

-   username: `admin`
-   password: `admin`

## Work with git

```{important}
This section applies to members of the GitHub `plone/developers` team, who have write access to repositories under the Plone GitHub organization.

Members of the `plone/contributors` team do not have write access, and instead must follow the process to set up their remote upstream and origin branches as described in {ref}`set-up-your-environment-label`.
```

Always begin by checking out the git branch on which you want to work.
If you have not yet checked out a branch, then you need to track it with the `-t` option and specify the remote to track, usually called `origin`.
The following command will switch and track changes from the remote `origin` and its branch `6.1`.

```shell

> git checkout -t origin/6.1

Reviewer: This is a bad example, because 6.1 is the default branch of buildout.coredev, so it's what you already have after cloning. It would probably make more sense to show how to switch to the 6.0 branch.

Author: Yeah, I tried to preserve something from 5.2, but it just doesn't make sense, now that we decided to not switch coredev branches. I rewrote this section in 26da8cc1d6fed9aa0adb12611384c29546c77e9e.

```

Now going forward, you need to specify only the local branch when you want to switch back to it from your development branch.

```shell
git checkout 6.1
```

Next pull down and merge any recent changes from the remote tracked repository with a single command.

```shell
git pull
```

## Edit packages

First identify the names of the Plone packages you want to work on.
If you do not know, you can ask in the [Plone Community Forum](https://community.plone.org/).
Only a few packages are in {file}`src/` by default.

Next create a new file {file}`buildout.local.cfg`, and add the names of packages that you want to develop under the `auto-checkout` list.

```cfg
[buildout]
extends =
  buildout.cfg

auto-checkout =
    # Add packages that you want to develop
    plone.app.event
    icalendar
    # others
    ...
```

When you make changes in your package, then rerun buildout with the following command, specifying your new buildout configuration file with the `-c` option.
You can add the `-N` flag to save time by not checking PyPI to see if there are updates to packages that were already installed.

```shell
./bin/buildout -c buildout.local.cfg -N
```

```{seealso}

> For tips on working with `mr.developer`, see {doc}`mrdeveloper`.

Reviewer: ```suggestion
Checking out additional repositories using the `auto-checkout` option is a feature of `mr.developer`.
For more information, see {doc}`mrdeveloper`.
```

Author: Done in a80c9ccc6fb197b567e6dcc8e62d3be5164d96e0

```

````{tip}
To avoid conflicts with `buildout.coredev` files, you can configure git for your user.
Either create or edit a file at {file}`~/.gitconfig`.
Then add the following stanza to it.

```cfg
[core]
    excludesfile = ~/.gitignore_global
```

Then add any standard `.gitignore` syntax to exclude files from getting committed and pushed to a remote repository.
````

Next create a new development branch on which you want to work from the current branch, tracking the upstream Plone repository, and check it out.
It's a good idea to use a branch name that includes the issue number and is descriptive of what it resolves.

```shell

> git switch -c 123-thing-i-fixed -t origin/6.1

Reviewer: ```suggestion
git switch -c 123-thing-i-fixed
```

`origin/6.1` only makes sense for buildout.coredev itself, which is not where we're creating this branch. And we don't want the new branch to track an existing remote branch -- the upstream branch it should track won't exist on the remote until it is pushed.

Author: Included with 26da8cc1d6fed9aa0adb12611384c29546c77e9e.

The tool that installs Plone core is `buildout.coredev`.

The current default and development branch of `buildout.coredev` is `6.1`
Older versions are named according to their `major.minor` version.
Its versions align with Plone's `major.minor` versions.

Use a separate directory for each version of Plone to which you want to contribute.
This will avoid switching between git branches, then re-running buildout, which can cause dependency conflicts between versions of Plone.

To set up a Plone 6 development environment, change your working directory to wherever you place your projects, and clone https://github.com/plone/buildout.coredev.
You can specify the branch that you want to check out with the `-b` option.

```shell
cd [MY_PROJECTS]
# clone a specific major.minor version branch
git clone -b 6.1 https://github.com/plone/buildout.coredev
cd buildout.coredev
```

````{important}
If you want to use a Python version that is not 3.11, follow these instructions.

Open the file {file}`bootstrap.sh` at the root of the repository.
Notice that the script expects Python 3.11 to be installed on your system and in your user's `PATH`.

```shell
#/bin/sh
`which python3.11` -m venv .
```

> Edit it according to the Python version you want to use, then save and close the file.

Reviewer 2: That will leave a local change.  Should we warn about it?

```suggestion
Edit it according to the Python version you want to use, then save and close the file.
After you have run the script you should undo the change, otherwise you have a local change in git that you might accidentally commit.
```

Author: A warning would be good, but it would be even better to allow the developer to use whatever version of Python they want as an argument in the script. I'm not sure how to do that with shell scripts.

Same with the `.bootstrap-pyenv.sh` script.

````

Now run the script to install Plone 6.

```shell
./bootstrap.sh
```

> If you run into issues in this process, see {doc}`troubleshooting`.

Reviewer 2: On the [preview](https://plone6--1671.org.readthedocs.build/contributing/core/index.html#install-plone-core-for-development) "troubleshooting" is no link: the `{doc}` does not seem to do anything.

Author: Fixed in be8c6ad660b7c14108d719ec5ce8d7f2f1512c27

git switch -c 123-thing-i-fixed
```

Now you can edit your code without affecting the original branch.

(contributing-core-test-locally-label)=

## Test locally

If you change the expected behavior of a feature in a package, you should write a test to cover the change.

To run a test for the specific package that you modified, use the `-m` option followed by the package name, as shown in the following example.

```shell
./bin/test -m plone.app.event
```

If any test fails, do not commit and push the changes.
Instead write a test that passes.

After the package level tests pass with your change, you can {ref}`contributing-core-create-a-pull-request-label` and let CI run and ask Jenkins to run the full test suite.

However, if CI or Jenkins report a test failure that you want to troubleshoot locally, you can run the full unit test suite to ensure other packages aren't affected by the change.
It takes 5-10 minutes to run the full unit test suite.

```shell
# Run unit tests
./bin/test
```

If you run acceptance tests with the `--all` option, it will repeatedly launch and close browser windows that gain focus, disrupting you from doing any other work.
This takes 30-40 minutes to run.

> If you can't afford this disruption, you can defer it to Jenkins.

Reviewer 2: This does not happen for me.  I have `export ROBOT_BROWSER="headlesschrome"` in my bash profile, and the Jenkins jobs have this too.  Maybe use something like this to explain it:

```suggestion
If you run acceptance tests with the `--all` option, it will run tests in a real browser.
This takes 30-40 minutes to run.
This may repeatedly launch and close browser windows that gain focus, disrupting you from doing any other work.
If this happens, you can install the `chromedriver` OS package.
See https://developer.chrome.com/docs/chromedriver.
Then run `export ROBOT_BROWSER="headlesschrome"` and again run `bin/test --all`.
```

Author: Nice! Thank you!

```shell
./bin/buildout -c buildout.local.cfg -N
```

```{seealso}
`mr.developer` checks out additional repositories using the `auto-checkout` option.
For more information, see {doc}`mrdeveloper`.
```

````{tip}
To avoid conflicts with `buildout.coredev` files, you can configure git for your user.
Either create or edit a file at {file}`~/.gitconfig`.
Then add the following stanza to it.

```cfg
[core]
    excludesfile = ~/.gitignore_global
```

Then add any standard `.gitignore` syntax to exclude files from getting committed and pushed to a remote repository.
````

Next create a new development branch on which you want to work from the current branch, tracking the upstream Plone repository, and check it out.
It's a good idea to use a branch name that includes the issue number and is descriptive of what it resolves.

```shell
git switch -c 123-thing-i-fixed
```

Now you can edit your code without affecting the original branch.

(contributing-core-test-locally-label)=

## Test locally

If you change the expected behavior of a feature in a package, you should write a test to cover the change.

> To run a test for the specific package that you modified, use the `-m` option followed by the package name, as shown in the following example.
> 
> ```shell
> ./bin/test -m plone.app.event

Reviewer 3: ```suggestion
To run a test for the specific package that you modified, use the `-s` option followed by the package name, as shown in the following example.

```shell
./bin/test -s plone.app.event
```

We should use `-s` as we are specifying a package, not a single module
